File: ruby/042_MultipleOf9.rb
# 042 - Multiple of 9（★4）
# dp, 動的計画法

MOD = 10 ** 9 + 7
K = gets.to_i

dp = [0] * (K + 1)
dp[0] = 1

# 正の整数Xが9の倍数で各桁の和がKになることと、
# Kが9の倍数になることとが同義
if K % 9 == 0
  (1..K).each do |i|
    (1..9).each do |j|
      dp[i] += dp[i - j] if i >= j
      dp[i] %= MOD
    end
  end
end

puts dp[K]

File: kotlin/042_MultipleOf9.kt
/**
 * 042 - Multiple of 9（★4）
 * dp, 動的計画法
 */

fun main() {
  val MOD = 1000_000_000 + 7
  val K = readLine()!!.toInt()
  val dp = MutableList(K + 1) { 0 }
  dp[0] = 1

  if (K % 9 == 0) {
    for (i in 1 .. K) {
      for (j in 1 .. 9) {
        if (i >= j) dp[i] += dp[i - j] % MOD
        dp[i] %= MOD
      }
    }
  }
  println(dp[K])
}
